### PR TITLE
Support for img norm in graph during onnx export

### PR DESCRIPTION
**Pull request by nick-clarifai**
## Motivation

The purpose of this PR is to provide an option that does the image norm step of inference within the graph itself during the export to an ONNX model.

#### `mmseg/datasets/pipelines/transforms.py`:
- an init flag is added which causes the image norm to be skipped during `__call__` and causes the flag to be passed along in the `img_norm_cfg`

#### `mmseg/models/segmentors/base.py `
- `forward` modified to do image normalization to the input if `img_norm_cfg['normalize_in_graph']` is True

#### `tools/pytorch2onnx.py`:

- command line arg added
- normalize step in test pipeline is modified to skip normalization but still add `img_norm_cfg`
- input is cloned in verification step so the pre-image norm input can be compared

## Use cases

This would allow one to have a totally contained model that doesn't require extra python code for inference so that the ONNX model could easily be used in a setting such as NVIDIA Triton.

### CODE REVIEW

**Comment by Junjun2016** [comment]
Hi @nick-clarifai 
Thanks for your contribution.
We will review it in the near future.

File: mmseg/models/segmentors/base.py [hunk 1]
```diff
@@ -104,6 +104,15 @@ def forward(self, img, img_metas, return_loss=True, **kwargs):
...
         the outer list indicating test time augmentations.
         """
+        if torch.onnx.is_in_onnx_export() or kwargs.pop('do_norm', False):
+            assert len(img_metas) == 1
+            img_norm_cfg = img_metas[0][0].get('img_norm_cfg', None)
+            if img_norm_cfg and img_norm_cfg['normalize_in_graph']:
+                mean = torch.tensor(img_norm_cfg['mean'])[None, ..., None,
+                                                          None]
+                std = torch.tensor(img_norm_cfg['std'])[None, ..., None, None]
+                img[0] = (img[0] - mean) / std
```
Comment (Junjun2016) [inline mmseg/models/segmentors/base.py:114]: L114
img[0] = (img[0] - mean) / std should be img = (img - mean) / std?

Comment (Junjun2016) [reply mmseg/models/segmentors/base.py:114]: If yes, it is better to move to the corresponding part.

**Comment by RockeyCoss** [comment]
Please merge the master branch into your branch, thank you.

File: mmseg/models/segmentors/base.py (hunk 1 above)
Comment (nick-clarifai) [reply mmseg/models/segmentors/base.py:114]: You're right that line should deal with each image in the list. Also it does happen to only be used when `return_loss=False` for now because it's only used in the onnx exporting script. I'll make both changes

**Comment by Junjun2016** [comment]
Hi @RunningLeon
Please review it.

File: mmseg/models/segmentors/base.py [hunk 2]
```diff
@@ -107,6 +107,16 @@ def forward(self, img, img_metas, return_loss=True, **kwargs):
...
         else:
+            if torch.onnx.is_in_onnx_export() or kwargs.pop('do_norm', False):
+                assert len(img_metas) == 1
+                img_norm_cfg = img_metas[0][0].get('img_norm_cfg', None)
+                if img_norm_cfg and img_norm_cfg['normalize_in_graph']:
+                    mean = torch.tensor(img_norm_cfg['mean'])[None, ..., None,
+                                                              None]
+                    std = torch.tensor(img_norm_cfg['std'])[None, ..., None,
+                                                            None]
+                    img = [(i - mean) / std for i in img]
```
Comment (RunningLeon) [inline mmseg/models/segmentors/base.py:118]: Since dynamic axis on batch dim is supported in pytorch2onnx in mmseg, could we do normalization with batch dim?

**Comment by RunningLeon** [comment]
@nick-clarifai Hi, thanks for your PR. If you add `--normalize-in-graph` in pytorch2onnx.py, then you might need to support it in downstream tool onnx2tensorrt.py as well.

File: mmseg/models/segmentors/base.py (hunk 2 above)
Comment (nick-clarifai) [reply mmseg/models/segmentors/base.py:118]: I'm sorry I don't quite understand what change you're proposing. I do know that the resulting model after normalizing in graph with `--normalize-in-graph` and `--dynamic-export` set does have a dynamic batch axis (assuming `cfg.model.test_cfg.mode == 'whole'`).

**Comment by nick-clarifai** [comment]
> @nick-clarifai Hi, thanks for your PR. If you add `--normalize-in-graph` in pytorch2onnx.py, then you might need to support it in downstream tool onnx2tensorrt.py as well.

Yeah this makes sense. The change I can see being necessary would be allowing image normalization in the [preprocessing step](https://github.com/open-mmlab/mmsegmentation/blob/master/tools/onnx2tensorrt.py#L26) to be skipped and adding a command line arg to skip normalization? I'm happy to make these changes if we're on the same page. Let me know if there's something else I'm forgetting there.

File: mmseg/models/segmentors/base.py (hunk 2 above)
Comment (RunningLeon) [reply mmseg/models/segmentors/base.py:118]: What I'm trying to say is that have you tested batch inference with batch_size>1 when set both `--normalize-in-graph` and `--dynamic-export` and if it is OK?

**Comment by RunningLeon** [comment]
> > @nick-clarifai Hi, thanks for your PR. If you add `--normalize-in-graph` in pytorch2onnx.py, then you might need to support it in downstream tool onnx2tensorrt.py as well.
> 
> Yeah this makes sense. The changes I can see being necessary would be allowing image normalization in the [preprocessing step](https://github.com/open-mmlab/mmsegmentation/blob/master/tools/onnx2tensorrt.py#L26) to be skipped and adding a command line arg to skip normalization? I'm happy to make these changes if we're on the same page. Let me know if there's something else I'm forgetting there.

Yes. `--normalize-in-graph` should be added to `onnx2tensorrt.py` as well.

File: mmseg/models/segmentors/base.py (hunk 2 above)
Comment (nick-clarifai) [reply mmseg/models/segmentors/base.py:118]: Ah, good point. Thanks for bringing it up. I just checked and ensured that batch inference with batch_size > 1 does work with models exported using `--normalize-in-graph` and `--dynamic-export`.

**Comment by nick-clarifai** [comment]
Thanks for the feedback. `onnx2tensorrt.py` now has an option to skip image normalization in the preprocessing step. I didn't think it made sense to name the arg `--normalize-in-graph` because the tensorrt model will only do so if the input onnx model does normalize in graph, so I named the arg `--skip-normalize` instead. Let me know if you suggest any more changes.

**Comment by RunningLeon** [comment]
> Thanks for the feedback. `onnx2tensorrt.py` now has an option to skip image normalization in the preprocessing step. I didn't think it made sense to name the arg `--normalize-in-graph` because the tensorrt model will only do so if the input onnx model does normalize in graph, so I named the arg `--skip-normalize` instead. Let me know if you suggest any more changes.

Actually, `skip-normalize` works for me.

**Comment by nick-clarifai** [comment]
Is there something else that should be done before this could be merged? Let me know if there's anything I can do.

File: tools/pytorch2onnx.py [hunk 3]
```diff
@@ -76,11 +76,15 @@ def _demo_mm_inputs(input_shape, num_classes):
...
-                       rescale_shape=None):
+                       rescale_shape=None,
+                       normalize_in_graph=False):
     # build the data pipeline
     if shape is not None:
         test_pipeline[1]['img_scale'] = (shape[1], shape[0])
     test_pipeline[1]['transforms'][0]['keep_ratio'] = False
+    for transform in test_pipeline[1]['transforms']:
+        if transform['type'] == 'Normalize':
+            transform.normalize_in_graph = normalize_in_graph
```
Comment (RunningLeon) [inline tools/pytorch2onnx.py:87]: Should explicitly convert input tensor from `uint8` to `float32`
by update line 93:
```
imgs = [img.float() for img in data['img']]
```

File: tools/onnx2tensorrt.py [hunk 4]
```diff
@@ -26,11 +26,15 @@ def get_GiB(x: int):
...
-                       rescale_shape: Optional[Iterable] = None) -> dict:
+                       rescale_shape: Optional[Iterable] = None,
+                       normalize_in_graph: Optional[bool] = False) -> dict:
     # build the data pipeline
     if shape is not None:
         test_pipeline[1]['img_scale'] = (shape[1], shape[0])
     test_pipeline[1]['transforms'][0]['keep_ratio'] = False
+    for transform in test_pipeline[1]['transforms']:
+        if transform['type'] == 'Normalize':
+            transform.normalize_in_graph = normalize_in_graph
```
Comment (RunningLeon) [inline tools/onnx2tensorrt.py:37]: Should explicitly convert input tensor from `uint8` to `float32`
by update line 43:
```
imgs = [img.float() for img in data['img']]

**Comment by RunningLeon** [comment]
> Is there something else that should be done before this could be merged? Let me know if there's anything I can do.

@nick-clarifai Hi, sorry for the late reply. After resolve the rest comments, this PR is ready to go merge for me.

**Comment by nick-clarifai** [comment]
> > Is there something else that should be done before this could be merged? Let me know if there's anything I can do.
> 
> @nick-clarifai Hi, sorry for the late reply. After resolve the rest comments, this PR is ready to go merge for me.

Thanks for the feedback @RunningLeon. I've made the requested changes.

**Review by RunningLeon — APPROVED** [review]
LGTM

**Review by Junjun2016 — APPROVED** [review]

**Comment by MengzhangLI** [comment]
Hi, @nick-clarifai nick, sorry for late reply.

This pr is closely to merge, could you please fix unit test problem? Your modification is not coveraged by unit test.

image

Best,

**Review by MengzhangLI — APPROVED** [review]

**Comment by nick-clarifai** [comment]
> Hi, @nick-clarifai nick, sorry for late reply.
> 
> This pr is closely to merge, could you please fix unit test problem? Your modification is not coveraged by unit test.
> 
> image
> 
> Best,

Sure, can you help me decide which is the best test to write? Should I write a test that runs the code from `pytorch2onnx.py` with the verification enabled? Or should I do something like add a test to `tests/test_models/test_segmentors/test_encoder_decoder.py` that makes sure the added code in `mmseg/models/segmentors/base.py` is run?

**Comment by OpenMMLab-Assistant005** [comment]
Hi @nick-clarifai ！We are grateful for your efforts in helping improve this open-source project during your personal time. 
Welcome to join OpenMMLab Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA
If you have a WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤
